### PR TITLE
Development bzx compare

### PR DESCRIPTION
**Pull request by tjcloa**
Updates from the bzx codebase mostly aimed at security issues fixes and optimization 
Closes issue #117

### CODE REVIEW

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 1]
```diff
@@ -391,7 +382,7 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
 		uint256 totalSupply = _totalAssetSupply(0);
 		uint256 totalBorrow = totalAssetBorrow();
 		if (totalSupply > totalBorrow) {
-			return totalSupply.sub(totalBorrow);
+			return totalSupply - totalBorrow;
```
Comment (computerphysicslab) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:391]: Avoiding safemath... there must be a good reason to do it. Any explanation?

Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:391]: `if (totalSupply > totalBorrow)`
guarantees there won't be overflow
using safe math here is redundant, just wasting gas

**Review by computerphysicslab — APPROVED** [review]

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 2]
```diff
@@ -285,14 +294,13 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
...
 		}
 
-		uint256 _balancesFrom = balances[_from];
-		require(_value <= _balancesFrom && _to != address(0), "14");
+		require(_to != address(0), "15");
 
-		uint256 _balancesFromNew = _balancesFrom.sub(_value);
+		uint256 _balancesFrom = balances[_from];
+		uint256 _balancesFromNew = _balancesFrom.sub(_value, "16");
 		balances[_from] = _balancesFromNew;
```
Comment (eMarchenko) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:304]: `balances[_from] = _balancesFrom.sub(_value, "16");`

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 3]
```diff
@@ -322,14 +330,13 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
 		uint256 _newBalance,
 		uint256 _currentPrice
 	) internal {
-		// keccak256("iToken_ProfitSoFar")
-		bytes32 slot = keccak256(abi.encodePacked(_user, uint256(0x37aa2b7d583612f016e4a4de4292cb015139b3d7762663d06a53964912ea2fb6)));
+		bytes32 slot = keccak256(abi.encodePacked(_user, iToken_ProfitSoFar));
```
Comment (eMarchenko) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:333]: check out [eip-1967](https://eips.ethereum.org/EIPS/eip-1967) approach to reduce collision chance even further

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 4]
```diff
@@ -932,6 +924,22 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
...
+		uint256 interestRate,
+		uint256 maxDuration,
+		uint256 marginAmount
+	) internal pure returns (uint256) {
+		return
+			maxDuration != 0
+				? interestRate
+					.mul(WEI_PERCENT_PRECISION)
+					.mul(maxDuration)
+					.div(31536000) // 86400 * 365
```
Comment (eMarchenko) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:937]: `.div(365 days)`

File: contracts/modules/LoanMaintenance.sol [hunk 5]
```diff
@@ -317,8 +317,8 @@ contract LoanMaintenance is LoanOpeningsEvents, LoanMaintenanceEvents, VaultCont
 	) external view returns (LoanReturnData[] memory loansData) {
 		EnumerableBytes32Set.Bytes32Set storage set = isLender ? lenderLoanSets[user] : borrowerLoanSets[user];
 
-		uint256 end = count.min256(set.values.length);
-		if (end == 0 || start >= end) {
+		uint256 end = start.add(count).min256(set.length());
+		if (start >= end) {
```
Comment (eMarchenko) [inline contracts/modules/LoanMaintenance.sol:321]: there are several instances of `for (uint256 i = end - start; i > 0; i--) {` loop. Consider writing it as 
`for (uint256 i = start; i < end; i++) {` if possible (i.e. it's ok to traverse items from start to end), as it is less error-prone

File: contracts/modules/SwapsExternal.sol [hunk 6]
```diff
@@ -42,7 +42,14 @@ contract SwapsExternal is VaultController, SwapsUser {
 			require(msg.value == sourceTokenAmount, "sourceTokenAmount mismatch");
 			wrbtcToken.deposit.value(sourceTokenAmount)();
 		} else {
+			IERC20 sourceTokenContract = IERC20(sourceToken);
+
+			uint256 balanceBefore = sourceTokenContract.balanceOf(address(this));
+
 			IERC20(sourceToken).safeTransferFrom(msg.sender, address(this), sourceTokenAmount);
+
+			// explicit balance check so that we can support deflationary tokens
```
Comment (eMarchenko) [inline contracts/modules/SwapsExternal.sol:51]: Deflationary token design is an anti-pattern.
Personally, I'd vote against their support as a way to discourage people implementing tokens that don't comply with ERC20 standard

**Review by eMarchenko — APPROVED** [review]

**Review by computerphysicslab — APPROVED** [review]

**Review by computerphysicslab — APPROVED** [review]

**Review by computerphysicslab — APPROVED** [review]

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol (hunk 2 above)
Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:304]: _balancesFromNew and _balancesToNew are used in L313-L314:
`_updateCheckpoints(_from, _balancesFrom, _balancesFromNew, _currentPrice);`
it is cheaper to read from a memory than from the storage, right?

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol (hunk 4 above)
Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:937]: agreed

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 7]
```diff
@@ -54,7 +60,6 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
         payable
         nonReentrant
         pausable(msg.sig)
-        settlesInterest
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:56]: best practice is not to modify storage inside modifier logic
`	modifier settlesInterest() {
		_settleInterest();
		_;
	}
`
removed the modifier as `_settleInterest();` is called directly throughout the code

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol (hunk 3 above)
Comment (ororopickpocket) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:328]: @tjcloa did you make sure that this does not change the storage slot currently used by the contracts? (just want to be safe here)

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 8]
```diff
@@ -341,9 +343,8 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
 
 	/* Public View functions */
 
-	function profitOf(address user) public view returns (uint256) {
-		// keccak256("iToken_ProfitSoFar")
-		bytes32 slot = keccak256(abi.encodePacked(user, uint256(0x37aa2b7d583612f016e4a4de4292cb015139b3d7762663d06a53964912ea2fb6)));
+	function profitOf(address user) public view returns (int256) {
```
Comment (ororopickpocket) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:346]: victor is using the function on the frontend and this changes the abi slightly. if you're changing any public functions, Victor needs to know about it.

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 9]
```diff
@@ -572,6 +558,7 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
 		require(burnAmount != 0, "19");
 
 		if (burnAmount > balanceOf(msg.sender)) {
+			require(burnAmount == uint256(-1), "32");
```
Comment (ororopickpocket) [inline contracts/connectors/loantoken/LoanTokenLogicStandard.sol:561]: why?

File: contracts/mixins/FeesHelper.sol [hunk 10]
```diff
@@ -16,12 +16,31 @@ contract FeesHelper is State, ProtocolTokenUser, FeesEvents {
...
-		return feeTokenAmount.mul(borrowingFeePercent).div(10**20);
+		return feeTokenAmount.mul(borrowingFeePercent).divCeil(10**20);
+		/*
+		// p3.9 from bzx peckshield-audit-report-bZxV2-v1.0rc1.pdf
+		// cannot be applied solely nor with LoanOpenings.sol as it drives to some other tests failure
+		uint256 collateralAmountRequired =
+			feeTokenAmount.mul(10**20).divCeil(
+				10**20 - borrowingFeePercent // never will overflow
+			);
+		return collateralAmountRequired.sub(feeTokenAmount);*/
```
Comment (ororopickpocket) [inline contracts/mixins/FeesHelper.sol:43]: why is this here?

Comment (tjcloa) [reply contracts/mixins/FeesHelper.sol:43]: postponed for the moment
https://github.com/DistributedCollective/Sovryn-smart-contracts/issues/190

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol [hunk 11]
```diff
@@ -932,6 +919,17 @@ contract LoanTokenLogicStandard is LoanTokenSettingsLowerAdmin {
...
+		uint256 interestRate,
+		uint256 maxDuration,
+		uint256 marginAmount
+	) internal pure returns (uint256) {
+		return
+			maxDuration != 0
+				? interestRate.mul(WEI_PERCENT_PRECISION).mul(maxDuration).div(365 days).div(marginAmount).add(WEI_PERCENT_PRECISION)
+				: WEI_PERCENT_PRECISION;
+	}
+
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:932]: will double check and remove it then

File: contracts/connectors/loantoken/LoanTokenLogicStandard.sol (hunk 3 above)
Comment (tjcloa) [reply contracts/connectors/loantoken/LoanTokenLogicStandard.sol:328]: double checked once again just in case, created test contract - the result is identical
